File: big-o/example12.py

def permutation(string):
    permuations(string, "")

def permuations(string, prefix):
    if len(string) == 0 :
        print(prefix)
    else:
        for i in range(len(string)):
            rem = string[0:i] + string[i+1::]
            permuations(rem, prefix + string[i])


permutation("abcd")

sammy = "sammy"

print(sammy[0:1])
print(range(len(sammy)))

File: big-o/example12.java
public class example12 {

    static void permutation(String str) 
    {
        permutation(str,"");
    }

    static void permutation(String str, String prefix)
    {
        if(str.length() == 0)
        {
            System.out.println(prefix);
        }
        else
        {
            for(int i = 0; i < str.length(); i++)
            {
                String rem = str.substring(0, i) + str.substring(i+1);
                System.out.println("recursion");
                permutation(rem, prefix + str.charAt(i));
            }
        }
    }

    public static void main(String[] args) throws Exception {
        System.out.println("test");
        permutation("abcd");
    }
}